nativelib/src/main/resources/gc/commix/Heap.c

```diff
@@ -0,0 +1,542 @@
+    } else {
+        return Heap_AllocSmall(heap, objectSize);
+    }
+}
+
+#ifdef DEBUG_ASSERT
+void Heap_clearIsSwept(Heap *heap) {
+    BlockMeta *current = (BlockMeta *)heap->blockMetaStart;
+    BlockMeta *limit = (BlockMeta *)heap->blockMetaEnd;
+    while (current < limit) {
+        BlockMeta *reserveFirst = (BlockMeta *) blockAllocator.reservedSuperblock;
+        BlockMeta *reserveLimit = reserveFirst + SWEEP_RESERVE_BLOCKS;
+        if (current < reserveFirst || current >= reserveLimit) {
+            assert(reserveFirst !=  NULL);
+            current->debugFlag = dbg_must_sweep;
+        }
+        current++;
+    }
+}
+
+void Heap_assertIsConsistent(Heap *heap) {
+    BlockMeta *current = (BlockMeta *)heap->blockMetaStart;
+    LineMeta *lineMetas = (LineMeta *)heap->lineMetaStart;
+    BlockMeta *limit = (BlockMeta *)heap->blockMetaEnd;
+    ObjectMeta *currentBlockStart = Bytemap_Get(heap->bytemap, heap->heapStart);
+    while (current < limit) {
+        assert(!BlockMeta_IsCoalesceMe(current));
+        assert(!BlockMeta_IsSuperblockStartMe(current));
+        assert(!BlockMeta_IsSuperblockTail(current));
+        assert(!BlockMeta_IsMarked(current));
+
+        int size = 1;
+        if (BlockMeta_IsSuperblockStart(current)) {
+            size = BlockMeta_SuperblockSize(current);
+        }
+        BlockMeta *next = current + size;
+        LineMeta *nextLineMetas = lineMetas + LINE_COUNT * size;
+        ObjectMeta *nextBlockStart =
+            currentBlockStart +
+            (WORDS_IN_BLOCK / ALLOCATION_ALIGNMENT_WORDS) * size;
+
+        for (LineMeta *line = lineMetas; line < nextLineMetas; line++) {
+            assert(!Line_IsMarked(line));
+        }
+        for (ObjectMeta *object = currentBlockStart; object < nextBlockStart;
+             object++) {
+            assert(!ObjectMeta_IsMarked(object));
+        }
+
+        current = next;
+        lineMetas = nextLineMetas;
+        currentBlockStart = nextBlockStart;
+    }
+    assert(current == limit);
+}
+#endif
+
+void Heap_Collect(Heap *heap) {
+#ifdef ENABLE_GC_STATS
+    Stats *stats = heap->stats;
+    if (stats != NULL) {
+        stats->collection_start_ns = scalanative_nano_time();
+    }
+#else
+    Stats *stats = NULL;
+#endif
+    assert(Sweeper_IsSweepDone(heap));
+#ifdef DEBUG_ASSERT
+    Heap_clearIsSwept(heap);
+    Heap_assertIsConsistent(heap);
+#endif
+    heap->mark.lastEnd_ns = heap->mark.currentEnd_ns;
+    heap->mark.currentStart_ns = scalanative_nano_time();
+    Marker_MarkRoots(heap, stats);
+    heap->gcThreads.phase = gc_mark;
+    // make sure the gc phase is propagated
+    atomic_thread_fence(memory_order_release);
+    // start main gc thread (0)
+    sem_post(&heap->gcThreads.start0);
+    while (!Marker_IsMarkDone(heap)) {
+        Marker_Mark(heap, stats);
+        if (!Marker_IsMarkDone(heap)) {
+            sched_yield();
+        }
+    }
+    heap->gcThreads.phase = gc_idle;
```

Comment:
This should also be a helper function. It would be nice to isolate all of the phase/thread management into its own module with clear method names.